### ISSUE TITLE
user-scheduler caught in loop fighting with default-scheduler

### ISSUE THREAD

**Issue by stevenstetzler** [jupyterhub/zero-to-jupyterhub-k8s#1592]
I am using the user-scheduler built into the helm chart to handle scheduling of user pods and user placeholder pods. I have my users and user placeholders scheduled onto nodes that have a specific label, right now it is something like `instance-type=t3-large`. (I intend to schedule users onto specific AWS EC2 instance types.)

If I try to schedule another pod that has the same node selector but use the Kubernetes default scheduler instead, the default scheduler will try to evict one of the `user-placeholder` pods and schedule my new pod onto the reserved node (1). This happens successfully, however it seems that the user-scheduler is blind to the existence of this new pod, and in some cases will act quickly to reschedule a `user-placeholder` onto the node that should now hold only my new pod before the new pod is actually scheduled (2). This then causes the default-scheduler to evict the `user-placeholder` pod again, returning to step (1), and a cycle ensues: (1)->(2)->(1)->(2)... I've found that this doesn't always happen, it seems that there is a race condition that results in a loop only some of the time (user-scheduler acts faster than default-scheduler or something).

I can fix this issue by assigning my new pod to be scheduled with the user-scheduler instead of the default-scheduler. This works fine in some cases, but in my specific application, I don't have the ability to change the scheduler for my new pod. The new pod in question is a Spark executor pod created with the [Spark on Kubernetes API](https://spark.apache.org/docs/latest/running-on-kubernetes.html). For older versions of Spark (all stable releases), the scheduler cannot be configured. For development version of Spark (~ 3.0), the pod template file can be edited like so to fix the issue:
```
apiVersion: v1
kind: Pod
metadata:
  name: spark
  labels:
    app: spark
spec:
  schedulerName: <namespace>-user-scheduler
```

I can also fix this issue by temporarily scaling the `user-scheduler` deployment to 0 in which case the default-scheduler has free reign to assign pods to nodes. Scaling the deployment back up to 2 after the new pod has been scheduled successfully results in expected behavior.

I am wondering if this race condition behavior is a problem with the implementation of the user-scheduler or if this behavior should be expected unless the new pod is scheduled with the user-scheduler.

Version Information:
- z2jh version: 0.9.0-beta.3
- Kubernetes version: EKS with Kubernetes version 1.14

**Comment by betatim** [comment]
Off the top of my head this sounds like a problem with the priorities of the pods. One thing to confirm is that the `user-scheduler` does not do anything to the spark pod and vice versa the `default-scheduler` should not interact with the placeholder pod. This is because the placeholder pod specifies that it should be scheduled by the user-scheduler and the spark pod doesn't specify a scheduler (so defaults to the default-scheduler).

The placeholder pod is configured to evict (itself or maybe it is its scheduler, not sure) when a pod of higher priority needs scheduling. The placeholder pods should have very low priority in order for all the "useful" pods to be bale to take their place. Did you configure a particular priority for your spark pod or a default priority for all pods in the cluster? Is that being applied correctly?

**Comment by consideRatio** [comment]
## Background theory refreshed
The user-scheduler is the official kubernetes `kube-scheduler` binary of a specific k8s version running with a custom configuration. In managed clusters (GKE at least), the default scheduler deployed in the kubernetes cluster is running out of sight, and may have a different version of the `kube-scheduler` binary, and is guaranteed to have a different configuration on what it considers most important for any pod it scheduled when evaluating what node it should schedule a pod to.

- The scheduler will allocate pending pods to run on a node, but only if it is assigned to schedule a specific pod as indicated by the pod's specification's `schedulerName`.
- The pod's specification's `priorityClassName` reference a priorityclass k8s resource, which simply declares a pod priority as a number like -10, 0, 1, or 10.
- Pods with a higher priority that are pending that doesn't fit unless a lower priority pod is evicted, will make the scheduler evict such lower priority pods.
- Pods with negative or more negative values than -10 will not make a cluster autoscaler add new nodes if they are found pending, this may be unrelated to the discussion...

## Feedback

> If I try to schedule another pod that has the same node selector but use the Kubernetes default scheduler instead, the default scheduler will try to evict one of the user-placeholder pods and schedule my new pod onto the reserved node (1).

So then I assume the user-placeholder pods had a lower priority through their referenced priorityclass. I would suggest confirming this by `kubectl get pod user-placeholder-0 -o yaml | grep priority`.

> This happens successfully, however it seems that the user-scheduler is blind to the existence of this new pod, and in some cases will act quickly to reschedule a user-placeholder onto the node that should now hold only my new pod before the new pod is actually scheduled (2). This then causes the default-scheduler to evict the user-placeholder pod again, returning to step (1), and a cycle ensues: (1)->(2)->(1)->(2)... I've found that this doesn't always happen, it seems that there is a race condition that results in a loop only some of the time (user-scheduler acts faster than default-scheduler or something).

Hmmm, are the user-placeholder pod evicting the pod that evicted the user-placeholder in the first place? That should not be able to happen as they cant both have higher priority than the other. Or... Hmm... Wait perhaps it can happen if a scheduler think one pod must be on a specific node while the other pod could be relocated to another node, but then that pod conclude it cannot be relocated to another node and must come back to the same node again, and now a vice versa situation.

---

Okay to continue considering this, I'd like to know the two competing pods `priority` field in their k8s specification as you get for example by using `kubectl get pod -n <namespace> <podname> -o yaml | grep priority`. Note that this priority is set dynamically and cannot be part of your initial definition, but indirectly only by specifying a priorityClassName referencing a priorityClass resource. We have created such for the user-placeholder pods to ensure they are lower than normal users instead of changing everything else to be of a higher priority.

**Comment by consideRatio** [comment]
I hope that this is no longer an issue after #1730 where I upgraded the kube-scheduler binary and so. Let's reopen this if this surface again for now. If so, please note the k8s version, and inspect the user-scheduler's image tag used with `kubectl describe deploy/user-scheduler | grep image`
